docs/developers/bindings.rst

```diff
@@ -0,0 +1,114 @@
+function defintion and nothing else.
+
+The next step is to convert the signature line into the prototype format that
+the script expects. For ``MPI_Send``, this should look something like this:
+
+.. code-block:: c
+
+    PROTOTYPE ERROR_CLASS send(BUFFER buf, COUNT count, DATATYPE type, RANK dest,
+                               TAG tag, COMM comm)
+
+Notice how the function name is changed, the ``MPI_`` prefix removed and the
+rest converted to lowercase, and also how each parameter is simplified into a
+``TYPE name`` format, where the ``TYPE`` conforms to an allowed list in
+``ompi/mpi/bindings/ompi_bindings/c_type.py``. For newer functions and types,
+you may have to extend the ``c_type.py`` file with a new class showing how to
+handle the type.
+
+The final step is to update ``Makefile.am``, adding the template name, in this
+case ``send.c.in``, to the ``prototype_sources`` variable, and the generated
+file name, ``generated_send.c``, to ``interface_profile_sources``. The
+generated file name must be of the form ``generated_${basename}.c``, where
+``${basename}`` is the name of the template file stripped of all extensions.
+
+Fortran Bindings
+----------------
+
+Adding a new Fortran binding may be as simple as adding an additional prototype
+to ``ompi/mpi/fortran/use-mpi-f08/interface.in``, as long as all the required
+types are already supported by the binding code. Below is an example prototype
+for ``MPI_Waitall``:
+
+.. code-block::
+
+    .waitall(SHORTCUT_COUNT count, REQUEST_ARRAY array_of_requests[count=count],
+             STATUS_ARRAY array_of_statuses[count=count])
+
+First, notice that the function name is listed with the ``MPI_`` prefix removed
+and in all lowercase along with a ``.`` before the name---this makes it easier
+to delimit prototypes across multiple lines. Parameters are listed in a
+simplified ``TYPE NAME`` form, with an optional ``[key=value;...]`` attribute
+coming after if required for the specific type. This key-value attribute is
+used to specify dependencies betwen parameters, where the key is validated by
+the type and the value must be the name of another parameter.
+
+The Fortran binding code not only generates Fortran, but also additional
+wrapping C code that calls into the C bindings, making conversions and checking
+for Fortran-specific error conditions as necessary. The following files will be
+generated by the script:
+
+* ``ompi/mpi/fortran/use-mpi-f08/api_f08_generated.F90``
+* ``ompi/mpi/fortran/use-mpi-f08/base/api_f08_generated.c``
+* ``ompi/mpi/fortran/use-mpi-f08/base/api_f08_ts_generated.c``
+* ``ompi/mpi/fortran/use-mpi-f08/mod/mpi-f08-interfaces-generated.h``
+
+The Fortran file ``api_f08_generated.F90`` contains all the internal subroutine
+definitions, each of which makes a call into corresponding C functions. Two
+different C files are generated: ``api_f08_ts_generated.c`` contains support
+for compilers with TS 29113 support, allowing the use of ``CFI_cdesc_t`` types;
```

Comment:
might be nice to add a hyperlink here, e.g. https://fortranwiki.org/fortran/show/Fortran+2018